zio-schema/shared/src/main/scala/zio/schema/DynamicValue.scala

```diff
@@ -110,12 +123,26 @@ sealed trait DynamicValue {
       case (dyn, Schema.Dynamic(_)) => Right(dyn)
 
       case _ =>
-        Left(() => s"Failed to cast $self to schema $schema")
+        Left(DynamicValue.Err.CastError(self, schema))
     }
 
 }
 
 object DynamicValue {
+
+  sealed trait Err
+
+  object Err {
+    final case class MissingCase(key: String, enumN: Schema.Enum[_]) extends Err
```

Comment:
Worth looking at the decoder / encoder errors to see if that's useful, applicable, or inspiational here.